### ISSUE TITLE
sql: expand `pg_table_is_visible` into an optimizable subquery

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#32939]
(requested by @jordanlewis)

**Is your feature request related to a problem? Please describe.**

Today `pg_table_is_visible()` is implemented as a built-in function evaluated again for every row. Internally the function runs a complex multi-join subquery with the given table name as input. This is super-slow.

For example `psql` when handling `\d` will run `pg_table_is_visible()` for every row of `pg_tables` and the cumulated work causes an unpleasantly long latency (bad UX).

**Describe the solution you'd like**

Substitute `pg_table_is_visible()` by a correlated subquery during scalar analysis (perhaps type checking? anyway, must be early) so that the resulting subquery can undergo simplification / decorrelation using the regular CBO machinery.

**Describe alternatives you've considered**

Alternative is to be cleverer and perhaps look up just the one table descriptor in `pg_table_is_visible()`. Problem with that is that builtins don't have access to the full SQL layer (and neither should they).

Jira issue: CRDB-4707

**Comment by knz** [comment]
@jordanlewis do we still want this?

**Comment by knz** [comment]
@mgartner I think we achieved this right? builtin functions that expand into SQL get merged in the overall plan?

**Comment by mgartner** [comment]
Correct! This was addressed in v23.1.

```
defaultdb> EXPLAIN (OPT) SELECT pg_table_is_visible(123);
                                          info
-----------------------------------------------------------------------------------------
  values
   └── tuple
        └── subquery
             └── project
                  ├── limit
                  │    ├── inner-join (lookup pg_namespace@pg_namespace_oid_idx [as=n])
                  │    │    ├── flags: force lookup join (into right side)
                  │    │    ├── scan pg_class@pg_class_oid_idx [as=c]
                  │    │    │    └── constraint: /3/2: [/123 - /123]
                  │    │    └── filters (true)
                  │    └── 1
                  └── projections
                       └── nspname IN ('pg_catalog', 'pg_extension', 'public')
```
